pkg/dm/spec/topology_dm.go

```diff
@@ -0,0 +1,742 @@
+func (s WorkerSpec) Status(masterList ...string) string {
+	if len(masterList) < 1 {
+		return "N/A"
+	}
+	masterapi := api.NewDMMasterClient(masterList, statusQueryTimeout, nil)
+	stage, err := masterapi.GetWorker(s.Name)
+	if err != nil {
+		return "Down"
+	}
+	if stage == "" {
+		return "N/A"
+	}
+	return stage
+}
+
+// Role returns the component role of the instance
+func (s WorkerSpec) Role() string {
+	return ComponentDMWorker
+}
+
+// SSH returns the host and SSH port of the instance
+func (s WorkerSpec) SSH() (string, int) {
+	return s.Host, s.SSHPort
+}
+
+// GetMainPort returns the main port of the instance
+func (s WorkerSpec) GetMainPort() int {
+	return s.Port
+}
+
+// IsImported returns if the node is imported from TiDB-Ansible
+func (s WorkerSpec) IsImported() bool {
+	return s.Imported
+}
+
+// PortalSpec represents the Portal topology specification in topology.yaml
+type PortalSpec struct {
+	Host            string                 `yaml:"host"`
+	SSHPort         int                    `yaml:"ssh_port,omitempty"`
+	Imported        bool                   `yaml:"imported,omitempty"`
+	Port            int                    `yaml:"port" default:"8280"`
+	DeployDir       string                 `yaml:"deploy_dir,omitempty"`
+	DataDir         string                 `yaml:"data_dir,omitempty"`
+	LogDir          string                 `yaml:"log_dir,omitempty"`
+	Timeout         int                    `yaml:"timeout" default:"5"`
+	NumaNode        string                 `yaml:"numa_node,omitempty"`
+	Config          map[string]interface{} `yaml:"config,omitempty"`
+	ResourceControl ResourceControl        `yaml:"resource_control,omitempty"`
+	Arch            string                 `yaml:"arch,omitempty"`
+	OS              string                 `yaml:"os,omitempty"`
+}
+
+// Role returns the component role of the instance
+func (s PortalSpec) Role() string {
+	return ComponentDMPortal
+}
+
+// SSH returns the host and SSH port of the instance
+func (s PortalSpec) SSH() (string, int) {
+	return s.Host, s.SSHPort
+}
+
+// GetMainPort returns the main port of the instance
+func (s PortalSpec) GetMainPort() int {
+	return s.Port
+}
+
+// IsImported returns if the node is imported from TiDB-Ansible
+func (s PortalSpec) IsImported() bool {
+	return s.Imported
+}
+
+// UnmarshalYAML sets default values when unmarshaling the topology file
+func (topo *DMTopologySpecification) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type topology DMTopologySpecification
+	if err := unmarshal((*topology)(topo)); err != nil {
+		return err
+	}
+
+	if err := defaults.Set(topo); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := fillDMCustomDefaults(&topo.GlobalOptions, topo); err != nil {
+		return err
+	}
+
+	return topo.Validate()
+}
+
+// platformConflictsDetect checks for conflicts in topology for different OS / Arch
+// for set to the same host / IP
+func (topo *DMTopologySpecification) platformConflictsDetect() error {
```

Comment:
Can we reuse the `cluster/spec` conflicts check?

Reply:
will do this in later pr, this file mainly copy from https://github.com/pingcap/tiup/pull/490/files
and fix many things not compile now...